File: frontend/src/types/features/assessment/question.ts
import type { QuestionOption, QuestionOptionRequest } from "./question-option";

export interface SimplifiedCategory {
  id: string;
  name: string;
}

export interface Question {
  id?: string;
  content: string;
  questionType: string;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  categoryId: string;
  options: QuestionOption[];
  tagIds: number[];
}

export interface QuestionListItem {
  id: string;
  content: string;
  questionType: string;
  isActive: boolean;
  categoryId: string;
  tagIds: number[];
  options: Omit<QuestionOption, 'id'>[];
}

export interface QuestionCreateRequest {
  content: string;
  questionType: string;
  isActive: boolean;
  categoryId: string;
  options: QuestionOptionRequest[];
  tagIds?: number[];
}

export interface QuestionFilter {
  categoryId?: string;
  tagIds?: number[];
  questionType?: string;
  isActive?: boolean;
  createdFrom?: string;
  createdTo?: string;
}

File: backend/src/main/java/com/example/starter_project_2025/system/assessment/question/Question.java
package com.example.starter_project_2025.system.assessment.question;

import com.example.starter_project_2025.system.assessment.question_category.QuestionCategory;
import com.example.starter_project_2025.system.assessment.question_option.QuestionOption;
import com.example.starter_project_2025.system.assessment.question_tag.QuestionTag;
import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.FieldDefaults;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.*;

@Entity
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Table(name = "questions")
@FieldDefaults(level = AccessLevel.PRIVATE)
public class Question {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    UUID id;

    @Column(columnDefinition = "TEXT", nullable = false)
    String content;

    @Enumerated(EnumType.STRING)
    QuestionType questionType;

    @ManyToOne
    @JoinColumn(name = "question_category_id")
    QuestionCategory category;

    @OneToMany(
            mappedBy = "question",
            cascade = CascadeType.ALL,
            orphanRemoval = true,
            fetch = FetchType.EAGER
    )
    List<QuestionOption> options = new ArrayList<>();

    @ManyToMany
    @JoinTable(
            name = "question_tags",
            joinColumns = @JoinColumn(name = "question_id"),
            inverseJoinColumns = @JoinColumn(name = "tag_id")
    )
    Set<QuestionTag> tags = new HashSet<>();

    Boolean isActive = true;

    @CreationTimestamp
    @Column(updatable = false)
    LocalDateTime createdAt;

    @UpdateTimestamp
    LocalDateTime updatedAt;
}